File: undirected_path.py
"""undirected_path.py"""
from collections import deque
from typing import Deque, List, Dict, Set

# Our simple undirected graph structure to play with
"""
┌───┐     ┌───┐
│   │     │   │
│ i ├─────┤ j │
│   │     │   │
└─┬─┘     └───┘
  │
  │
  │
  │
  │
┌─┴─┐     ┌───┐
│   │     │   │
│ k ├─────┤ l │
│   │     │   │
└─┬─┘     └───┘
  │
  │
  │
  │
  │
┌─┴─┐
│   │
│ m │
│   │
└───┘





┌───┐     ┌───┐
│   │     │   │
│ o ├─────┤ n │
│   │     │   │
└───┘     └───┘
"""

# Edge list
EDGES: List[List[str]] = [
    ["i", "j"],
    ["k", "i"],
    ["m", "k"],
    ["k", "l"],
    ["o", "n"],
]

# Convert to adjacency list, used to test against
GRAPH: Dict[str, List[str]] = {
    "i": ["j", "k"],
    "j": ["i"],
    "k": ["i", "m", "l"],
    "m": ["k"],
    "l": ["k"],
    "o": ["n"],
    "n": ["o"],
}


# Helper function
def _build_graph(edges: List[List[str]]) -> Dict[str, List[str]]:
    """Helper function to build undirected graph from edges."""
    graph: Dict[str, List[str]] = dict()
    _a: str
    _b: str
    for edge in edges:
        _a, _b = edge
        if _a not in graph:
            graph[_a] = list()
        if _b not in graph:
            graph[_b] = list()
        graph[_a].append(_b)
        graph[_b].append(_a)
    return graph


# Depth first has-path
def depth_first_has_path_recursive(graph: Dict[str, List[str]], src: str,
                                   dst: str, visited: Set[str]) -> bool:
    """Depth first has-path recursive algo with cyclical checks."""
    if src == dst:
        return True
    if src in visited:
        return False
    visited.add(src)
    for neighbor in graph[src]:
        if depth_first_has_path_recursive(graph, neighbor, dst, visited):
            return True
    return False


# Breadth first has-path
def breadth_first_has_path(graph: Dict[str, List[str]], src: str, dst: str,
                           visited: Set[str]) -> bool:
    """Breadth first has-path iterative algo with cyclical checks."""
    queue: Deque[str] = deque([src])
    while queue:
        current: str = queue.popleft()
        if current == dst:
            return True
        if current in visited:
            continue
        visited.add(current)
        for neighbor in graph[current]:
            queue.append(neighbor)
    return False


# Undirected graph traversal algos
def undirected_path_depth_first(edges: List[List[str]], node_a: str,
                                node_b: str) -> bool:
    """Undirected depth first has-path recursive algo
    with cyclical checks from edge list input."""
    graph: Dict[str, List[str]] = _build_graph(edges)
    return depth_first_has_path_recursive(graph, node_a, node_b, set())


def undirected_path_breadth_first(edges: List[List[str]], node_a: str,
                                  node_b: str) -> bool:
    """Undirected breadth first has-path iterative algo
    with cyclical checks from edge list input."""
    graph: Dict[str, List[str]] = _build_graph(edges)
    return breadth_first_has_path(graph, node_a, node_b, set())


# TESTS
TEST_GRAPH: Dict[str, List[str]] = _build_graph(EDGES)
assert TEST_GRAPH == GRAPH

assert undirected_path_depth_first(EDGES, "j", "m")
assert undirected_path_depth_first(EDGES, "j", "n") is False

assert undirected_path_breadth_first(EDGES, "j", "m")
assert undirected_path_breadth_first(EDGES, "j", "n") is False

File: rust_solutions/src/undirected_path.rs
use crate::graph::Graph;
use std::collections::{HashSet, VecDeque};

#[allow(dead_code)]
trait Dfs {
    fn depth_first_has_path_recursive(
        &self,
        src: &str,
        dst: &str,
        visited: &mut HashSet<String>,
    ) -> bool;
    fn undirected_path_depth_first(&self, node_a: &str, node_b: &str) -> bool;
}

#[allow(dead_code)]
trait Bfs {
    fn breadth_first_has_path(&self, src: &str, dst: &str, visited: &mut HashSet<String>) -> bool;
    fn undirected_path_breadth_first(&self, node_a: &str, node_b: &str) -> bool;
}

impl Dfs for Graph {
    fn depth_first_has_path_recursive(
        &self,
        src: &str,
        dst: &str,
        visited: &mut HashSet<String>,
    ) -> bool {
        if src == dst {
            return true;
        }
        if visited.contains(src) {
            return false;
        }
        visited.insert(src.to_string());
        if let Some(neighbors) = self.nodes.get(src) {
            for neighbor in neighbors {
                if Some(self.depth_first_has_path_recursive(neighbor, dst, visited)) == Some(true) {
                    return true;
                }
            }
        }
        false
    }

    fn undirected_path_depth_first(&self, node_a: &str, node_b: &str) -> bool {
        //Undirected depth first has-path recursive algo
        //with cyclical checks from edge list input.

        let mut visited: HashSet<String> = HashSet::new();
        self.depth_first_has_path_recursive(node_a, node_b, &mut visited)
    }
}

impl Bfs for Graph {
    fn breadth_first_has_path(&self, src: &str, dst: &str, visited: &mut HashSet<String>) -> bool {
        let mut queue: VecDeque<String> = VecDeque::new();
        queue.push_back(src.to_string());
        while let Some(current) = queue.pop_front() {
            if current == dst {
                return true;
            }
            if visited.contains(&current) {
                continue;
            }
            visited.insert(current.to_string());
            if let Some(neighbors) = self.nodes.get(&current) {
                for neighbor in neighbors {
                    queue.push_back(neighbor.clone());
                }
            }
        }
        false
    }

    fn undirected_path_breadth_first(&self, node_a: &str, node_b: &str) -> bool {
        //Undirected breadth first has-path iterative algo
        //with cyclical checks from edge list input.

        let mut visited: HashSet<String> = HashSet::new();
        self.breadth_first_has_path(node_a, node_b, &mut visited)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    // Our simple undirected graph structure to play with
    //
    // ┌───┐     ┌───┐
    // │   │     │   │
    // │ i ├─────┤ j │
    // │   │     │   │
    // └─┬─┘     └───┘
    //   │
    //   │
    //   │
    //   │
    //   │
    // ┌─┴─┐     ┌───┐
    // │   │     │   │
    // │ k ├─────┤ l │
    // │   │     │   │
    // └─┬─┘     └───┘
    //   │
    //   │
    //   │
    //   │
    //   │
    // ┌─┴─┐
    // │   │
    // │ m │
    // │   │
    // └───┘
    //
    //
    //
    //
    //
    // ┌───┐     ┌───┐
    // │   │     │   │
    // │ o ├─────┤ n │
    // │   │     │   │
    // └───┘     └───┘
    //

    fn create_graph() -> Graph {
        let mut graph = Graph::new();
        graph.add_edge(&["i", "j"]);
        graph.add_edge(&["k", "i"]);
        graph.add_edge(&["m", "k"]);
        graph.add_edge(&["k", "l"]);
        graph.add_edge(&["o", "n"]);
        graph.build_graph_from_edges();
        graph
    }

    #[test]
    fn depth_first_shortest_path_test() {
        let graph: Graph = create_graph();
        assert!(graph.undirected_path_depth_first("j", "m"));
        assert!(!graph.undirected_path_depth_first("j", "n"));
    }

    #[test]
    fn breadth_shortest_path_test() {
        let graph: Graph = create_graph();
        assert!(graph.undirected_path_breadth_first("j", "m"));
        assert!(!graph.undirected_path_breadth_first("j", "n"));
    }
}
